### PR TITLE
CORDA-3901 Reduce time taken for reconnect tests

### PR DESCRIPTION
**Pull request by rnicoll**
Reduce time taken for RPC reconnect tests, as well as ensuring the test is always applied
the same way on each run to avoid risk of an unstable test.

Also corrects a misuse of `AtomicInteger`, which read then separately wrote back to the container. Atomic integers are not locked, and this could have a potential race condition. Instead the code now collects every value, and finds the maximum at the end.

### CODE REVIEW

File: node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt [hunk 1]
```diff
@@ -76,10 +81,8 @@ class RpcReconnectTests {
      * Also the RPC connection is made through a proxy that introduces random latencies and is also periodically killed.
      */
     @Suppress("ComplexMethod")
-    @Test(timeout=420_000)
+    @Test(timeout=260_000)
```
Comment (blsemo) [inline node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt]: Why is the timeout for this test shorter than any bog standard unit test?

File: node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt [hunk 2]
```diff
@@ -169,15 +172,14 @@ class RpcReconnectTests {
...
                     if (flowsCountdownLatch.count == 0L) break
+                    val nodeRunningTime = Duration.ofMillis((randomSource.nextInt(nodeRunningTimeVariability) + nodeRunningTimeLowerBound).toLong())
 
                     // Let the node run for a random time interval.
-                    nodeRunningTime().also { ms ->
-                        log.info("Running node for ${ms / 1000} s.")
-                        Thread.sleep(ms.toLong())
-                    }
+                    log.info("Running node for ${nodeRunningTime.seconds} s.")
+                    Thread.sleep(nodeRunningTime.toMillis())
```
Comment (blsemo) [inline node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt]: What is the impact of sleeping for a shorter time here (on average)?

**Review by blsemo — COMMENTED** [review]
+1 for fixing the (ab)use of atomic int and making the code cleaner.
However, otherwise this is just tightening the parameters on a stochastic test - is it understood what the impact on coverage and test behaviour is?
Other question - does this test _need_ to be stochastic?

File: node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt (hunk 1 above)
Comment (rnicoll) [reply node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt]: The 300 seconds timeout appears to be a baseline we use unless we have well assessed data. As we know this ran reliably in 420 seconds previously, and we have approximately halved the workload, it seems reasonable to halve the timeout.

File: node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt (hunk 2 above)
Comment (rnicoll) [reply node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt]: Nodes are killed/interrupted faster, making for a more stressy test.

Comment (blsemo) [reply node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt]: Sure - but are we sure they get to the relevant point having started flows before getting killed?

**Comment by blsemo** [comment]
Run tests: slow

**Comment by rnicoll** [comment]
The test probably doesn't have to be stochastic, although honestly I'm fairly certain we don't need this test, and that `RPCStabilityTests` does a decent job of covering this. The idea here is to make a quick relatively safe improvement.

Resetting the node duration parameters in the next patch.

File: node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt (hunk 2 above)
Comment (rnicoll) [reply node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt]: Reset the timings for now, especially as they didn't seem to do a lot anyway.

**Comment by rnicoll** [comment]
Run tests: slow

**Comment by dimosr** [comment]
The `RpcStabilityTests` test the plain rpc client, while this test exercises the reconnecting rpc client.
However, we also have the `CordaRPCClientReconnectionTest` that verifies the behaviour of all the public APIs of the reconnecting rpc client in a more isolated way. 

This one was more of a pre-existing "chaos-testing" scenario that we kept around to be on the safe side. It's important to note it only exercises an API that's not public (`runFlowWithLogicalRetry`), but it's recommended in our docs to be used as a pattern for clients. Given the grief it has given us, I think it'd be a good idea to make it deterministic and shorter (e.g. exercise one failure category once or twice and always in the same order with the same delay between failures) since we're at it - I believe it shouldn't be that hard.

**Comment by rick-r3** [comment]
@fowlerrr was this one of the tests you had been looking at ?

File: node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt (hunk 1 above)
Comment (fowlerrr) [reply node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt]: That reliability only works for OS.  This test can take significantly longer on ENT.  Which is why you should find the timeout quite large there. (15 minutes).

In an ideal world this test would be excised and left for dead.  But it has some stress test value.  In accordance, the last time I suggested dropping the flow count to 50 I was suggested not to.

File: node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt (hunk 2 above)
Comment (fowlerrr) [reply node/src/integration-test-slow/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt]: There wasn't really any rhyme or reason to when they were killed before either.

**Review by fowlerrr — COMMENTED** [review]
At one point I did try to make this an actual deterministic test but gave up.  It's worth trying these changes in ENT before merging to see what effect they will have.  (I don't know why that has a specific impact, but it does)

Also, if possible, could we move this test to a separate workstream specifically for stress tests? If we have enough tests for that to make sense, of course.

**Comment by rnicoll** [comment]
@fenryka Suggest we just accept this isn't worth the effort and cut it, thoughts?

**Comment by fenryka** [comment]
Kill it! We have far too many tests that apparently "add value" yet that are flakey and detrimental to the overall health of our builds. If it's really that valuable functionality here, someone re-write it as part of a regression suite

**Review by fenryka — APPROVED** [review]
Kill it with fire

**Pull request merged**